### DISCUSSION TITLE
Can see server on all local devices except for Nintendo Switch

### DISCUSSION THREAD

**Discussion by PronkMedia** [itzg/docker-minecraft-bedrock-server#395] [Q&A]
First of all, not sure if this is the right place to ask/report this. If not, please close and tell me where to ask this.

I was able to setup a server in my local network, using docker compose with the default/example settings.
All devices in my network seem to be able to see the server and connect to it, except for my Nintendo Switch.

Now I am in doubt, is this perhaps just not supported? Could it be a configuration error? Am I missing additional configuration? Or something wrong in my local network?

**Comment by mattholy** [comment]
Nintendo Switch does NOT support third-part servers.
Not only Nintendo but also every game console.

**Comment by PronkMedia** [comment]
It's a LAN server, that's different than a third-party server, right? Or am I wrong here. 

My Android devices are connecting/playing without any problem on my server. So I don't see why the LAN server should not show up on my friend list on the Switch just like how it does on the other devices.

**Comment by mattholy** [comment]
A LAN server is of course a third-party server.
Maybe you could talk to Nintendo

**Comment by PronkMedia** [comment]
So, what are you saying then exactly. Will my server not work on Playstation, Switch and Xbox devices and only on Android? What devices are supported then for LAN play with a bedrock server? Can I find a list somewhere?

**Comment by DominusExult** [comment]
It doesn't work for Switch or rather Nintendo doesn't allow it (and the Switch is also stuck on a way older bedrock version). 
For sure it works with Windows, Android and iOS. I'd be surprised if it doesn't work for Xbox and I guess it will work for Playstation (unless Sony disalliws 3rd party servers, too).

**Comment by DominusExult** [comment]
I stand corrected, a quick search shows that consoles are generally not going to be able to coonect to 3rd party seevers.
There might be some workarounds, though.

**Comment by PronkMedia** [comment]
Thanks for the reactions/clarification on this. I was just not expecting it to not work when there is actually Bedrock server available for self hosting. 

I was able to implement a workaround in the meantime, that involves my own DNS server. As this seems to be a big issue for the community, I'm considering creating my own docker image that includes a DNS server with pre-defined settings for this. Maybe it's even an idea to include it (as optional setting) in this image. It would prevent the hassle of having to setup and configure additional third party apps.

No need to keep this open, so I will close this issue.

**Comment by DominusExult** [comment]
Yes, keep us posted on this. And I agree that adding this to this image could be beneficial!

**Comment by itzg** [comment]
To be proper the DNS server should remain separate image, **but** could be a "sidecar" container to this one. So, if you end up going that way, would be glad to include a link to that from the README here.

**Comment by josephdpurcell** [comment]
I've gone down the rabbit hole of trying to get a PS5 to be able to connect to a `docker-minecraft-bedrock-server`.

I see this issue is for Nintendo Switch, but its generally related to the category of "consoles".

What I see is this: "Due to platform restrictions, the external server feature is not available on consoles." (Source: [help.minecraft.net 4410316619533](https://help.minecraft.net/hc/en-us/articles/4410316619533-How-to-Play-Minecraft-Multiplayer#h_01FGSN86E5S5VJTWCSWJ3JRAG2))

To get around this limitation there are several options:

1. **LAN Games**: Ensure the console (Switch, PS5, etc) are on the same LAN. Under "Friends" tab a new section will display called "LAN Games"
2. **Proxy**: Use tunneling / proxying at the network layer to "fake" a LAN connection so that the server shows up under "LAN Games". There are several projects that I consider falling under this category:
    * [jhead/phantom](https://github.com/jhead/phantom) GitHub project. On the same LAN as your console you run this command with the server's IP/port as an argument and then that server gets bound to look like its on your LAN.
    * You can also just use a SSH tunnel or any other type of tunneling ([example 1 on StackExchange](https://superuser.com/questions/763388/proxying-minecraft-through-a-ssh-tunnel), [example 2 on Linode](https://www.linode.com/community/questions/18936/ssh-tunneling-for-minecraft-proxy)).
    * **BedrockTogher**: See https://bedrocktogether.net/. You download the app on an Android or iOS device, put in a port, and then the server will show up under Friends -> "LAN Games". I assume this is just a proxy/tunnel that binds a remote server so it looks like its LAN?
3. **VPN**: This would work similar to the Proxy option but with different specifics. The VPN destination network would need be have the Minecraft server on the same LAN.
4. **DNS Hijacking**: There is a class of solutions that involves pointing your console's DNS to your a DNS server that hijacks the DNS record for the featured server. The user joins a feature server which then joins a MC world which then prompts the user to enter a server IP/port to join. This is a fascinating approach. Just set your console's DNS to 104.238.130.180 (US), see more details on https://github.com/Pugmatt/BedrockConnect. I'm sure there are plenty of alternate ways to do this if you run your own DNS server
5. **Joinable Friends**: To do this, add a friend and then under "Friends" tab you should see a section "Joinable Friends" with the friend listed. Now, how do you get a friend to show up there? I don't know. The [Geyser docs](https://wiki.geysermc.org/geyser/using-geyser-with-consoles/) suggest you need [MCXboxBroadcast](https://github.com/rtm516/MCXboxBroadcast) or [FriendConnect](https://github.com/Androecia/FriendConnect). I don't actually understand how this works.

Questions I have:

1. Does Geyser do something special that allows a `docker-minecraft-bedrock-server` instance to be visible to a console? Based on the documentation https://geysermc.org/ it seems like no.
2. How does "Joinable Friends" actually work? In my testing, I added users on device A and device B as friends, A is LAN to the MC server and I logged in, but then B couldn't see A as a joinable friend. Has anyone had success with that?
3. The "Invite Friend" button is greyed out. I'm curious if this is its own class of solutions to getting a user on a console to be able to play on a `docker-minecraft-bedrock-server`? If so, how do you get the button to no longer be greyed out? I did some research about age and privacy restrictions on your Xbox account but didn't have luck.

PS. I can move this to a discussion if that makes more sense.

**Reply by josephdpurcell** [reply]
I am sorely mistaken. I need to revise my comment/research.

After doing some testing it seems like a console can only join a game if it's a Realm, meaning a server hosted/owned by Microsoft.

Third party / self hosted servers are not supported by any console (PS,Xbox,etc). That's what I'm seeing.

And, that's the conclusion others have come to as well.

The DNS hijacking appears to be an option, also perhaps this Joinable Friends thing is an option but it wouldn't surprise me if that only works for Realms. I wonder if the Joinable Friends solutions are really just to allow people to join a friend's Realm server when they aren't logged into it.

Ahhhhhhhhhhhhhhh Microsoft!!!!!

**Comment by itzg** [comment]
Thanks for the research and info @josephdpurcell 

(Good point about discussion. Let me convert this official into a discussion so it can be open)
